Added user scripts loading from addme file. experimental

milehighguy-dev opened the pull request:
New way of loading user scripts. Added functionality for defining the paths to arbitrary lvl files in the addme. This list then gets sent to the ingame mode and each level is loaded. The paths are saved in the mission setup table. I don't know how large this table can grow without problems.

milehighguy-dev committed:
Added user scripts loading from addme file. experimental

milehighguy-dev committed:
added call to DoFile with the trimmed filename. script name should match file name for it to run

milehighguy-dev committed:
added comments

milehighguy-dev committed:
made trim function also remove the file type from the string

BAD-AL commented on src/scripts/addme.lua:
It would be cool if there was a table that the mod creator could attach to the missionlist entry that gets passed to the ingame lua env.
```LUA
local newEntry = { mapluafile = "bes2%s_%s", 
                          era_g = 1, era_c = 1, 
                          extra_data= { my_special_data = "user_option:G2,H7,J5"},
                          mode_con_c = 1, mode_con_g = 1, mode_ctf_c = 1, mode_ctf_g = 1, mode_eli_g = 1,}

table.insert(sp_missionselect_listbox_contents , newEntry)
```
This way we can pass options from the shell/menu space over to the ingame space.
Perhaps just put 'extra_data' in the global env.
I believe this technique could be used to circumvent the addon mission limit too.

milehighguy-dev replied:
I could do that, but as far as I am aware, sp_missionselect_listbox_contents is only present in the shell, not ingame. Is that true?

And that would be a different pull request

milehighguy-dev replied:
I just tested it and sp_missionselect_listbox_contents is nil ingame

BAD-AL replied:
correct, the idea being that the 'extra_data' would get passed to the ingame lua env; not the entire 'sp_missionselect_listbox_contents'.
Might be good to put the actual mission name in the global ingame lua env too.
But it might be something to discuss with voice as it can be easier to convey ideas that way.

milehighguy-dev replied:
I don't think you can easily get what mission you are in while already in the mission. And I'm afraid i probably won't be able to do voice soon. Also this isn't quite relevant to this particular pull request, but we can discuss it on discord or make an issue on this repo

milehighguy-dev committed:
moved mission setup save to addme file, override EnterMission

milehighguy-dev commented on src/scripts/patch_ingame.lua:
```diff
@@ -18,23 +18,78 @@ ScriptCB_DoFile("utility_functions2")
+                    --we don't really have to do this but i figure its a good idea
+                    missionSetup.userScripts = nil
+                    ScriptCB_SaveMissionSetup(missionSetup)
+                end
```
I might remove this part because I'm not sure if you quit back to shell from ingame, do the addon files get loaded again? if they don't then the user script table will be nil

milehighguy-dev replied:
Nevermind, when you quit back to shell from a game, it goes through all the addme scripts again

BAD-AL commented on src/scripts/patch_ingame.lua:
```diff
@@ -18,23 +18,78 @@ ScriptCB_DoFile("utility_functions2")
-            ReadDataFile(patchDir .. "user_script_" .. i .. ".lvl")
-            ScriptCB_DoFile("user_script_" .. i)
+
+    -- retrieve user scripts we saved in mission setup table
```
Revisiting this functionality, I do think we need to keep the 1.3 uop user script stuff.
But adding functionality like this will be really useful.
I would like to see the ability to select if the script gets run based on mission name also.

```LUA
-- path - the path to the script lvl file
-- name_of_script_to_execute - the name of the script inside of the script lvl to run
-- list_of_applicable_mission_names - list of mission names the script applies to; nil if it applies to everything.
function AddUserScript(path, name_of_script_to_execute, list_of_applicable_mission_names)

end
--- -- OR -----
 -- path - the path to the script lvl file
-- name_of_script_to_execute - the name of the script inside of the script lvl to run
 -- bool_function - gets passed the mission name, returns true if it applies to the mission passed in
function AddUserScript(path, name_of_script_to_execute, bool_function)

end
```

The mission names get set inside the missionselect screen function 'ifs_missionselect_fnLaunch'
So we may want to override 'ScriptCB_SetMissionNames()' to apply the scripts to run
```LUA
function ifs_missionselect_fnLaunch(this)
	-- First entry is always "remove all", so we need to have more entries than
	-- that in the list.
	if(table.getn(gPickedMapList) > 1) then
		this.SelectedMap = 1
		ifs_missionselect_fnDeleteMap(this, 1) -- delete the "remove all entry"
		ScriptCB_SetMissionNames(gPickedMapList,ifs_instant_options_GetRandomizePlaylist())
		-- TODO: Maybe make the 'this.fnDone()'' call work?
		ScriptCB_SetTeamNames(0,0)
		ScriptCB_EnterMission()
		
		--this.fnDone()
	else
		ifelm_shellscreen_fnPlaySound(this.errorSound)
	end
end
```

milehighguy-dev replied:
You're right I should put the 1.3 way of adding user scripts back in too.

As for setting a user script to run based on that list, we still need a way to determine what mission is currently being played (during game time) if we want that to work

BAD-AL replied:
After further discussion, I do agree that the decision to run the script functionality is not necessary.
But I would like to see a 'remove' function as well (for use in menus, not so much in addme scripts).

```LUA
-- path - the path to the script lvl file
-- name_of_script_to_execute - the name of the script inside of the script lvl to run
function AddUserScript(path, name_of_script_to_execute )end

-- path - the path to the script lvl file
-- name_of_script_to_execute - the name of the script inside of the script lvl to run
function RemoveUserScript(path, name_of_script_to_execute )end
```

BAD-AL committed:
Merge remote-tracking branch 'origin/master' into mhg-userScript

BAD-AL committed:
1. Added 'RemoveUserScript' function
2. Fixed some issues with 'RunUserScripts()'
3. Verified system operation with debugger.
